lib/helpers.py

```diff
@@ -10,437 +10,598 @@
+
+        return merged_df[merged_df.ts_x > merged_df.ts_y]
+
+
+class _CSVHelpers(object):
+    def __init__(self, customer, revenue_event, chunk_size=10 ** 6):
+        """
+        Internal class, containing technical read-write related methods and helpers
+        :param customer: Name of the customer the report is created for
+        :param revenue_event: An event which is going to be taken as a revenue event, e.g. "purchase"
+        :param chunk_size: How many lines should be taken for a single chunk during reads
+
+        :type customer: str
+        :type revenue_event: str
+        :type chunk_size: int
+        """
+        self.customer = customer
+
+        self.revenue_event = revenue_event
+
+        self.chunk_size = chunk_size
+
+        # columns to load from CSV
+        self.columns = dict()
+        self.columns[CSV_SOURCE_MARKS_AND_SPEND] = ['ts', 'user_id', 'ab_test_group', 'campaign_id', 'cost_eur',
+                                                    'event_type']
+        self.columns[CSV_SOURCE_ATTRIBUTIONS] = ['ts', 'user_id', 'partner_event', 'revenue_eur']
+
+    def read_csv(self, audience, source, date, chunk_filter_fn=None):
+        """
+        Helper to download CSV files, convert to DF and print time needed.
+        Caches files locally and on S3 to be reused.
+        """
+        now = datetime.now()
+
+        date_str = date.strftime('%Y%m%d')
+
+        filename = '{0}/{1}/{2}.csv.gz'.format(self._audience_data_path(audience), source, date_str)
+
+        # local cache
+        cache_dir = '{0}/{1}/{2}'.format(cache_folder, audience, source)
+        if not os.path.exists(cache_dir):
+            os.makedirs(cache_dir)
+
+        cache_file_name = '{0}/{1}.parquet'.format(cache_dir, date_str)
 
-        # remove events without a user id
-        filtered_chunk = filtered_chunk[filtered_chunk['user_id'].str.len() == 36]
+        # s3 cache (useful if we don't have enough space on the Colab instance)
+        s3_cache_file_name = '{0}/{1}/{2}/{3}.parquet'.format(self._audience_data_path(audience), source, cache_folder,
+                                                              date_str)
 
-        filtered_chunk = improve_types(filtered_chunk)
+        if source == CSV_SOURCE_ATTRIBUTIONS:
+            cache_file_name = '{0}/{1}-{2}.parquet'.format(cache_dir, date_str, self.revenue_event)
 
-        df = pd.concat([df, filtered_chunk],
-                       ignore_index=True, verify_integrity=True)
+            # s3 cache (useful if we don't have enough space on the Colab instance)
+            s3_cache_file_name = '{0}/{1}/{2}/{3}-{4}.parquet' \
+                .format(self._audience_data_path(audience), source, cache_folder, date_str, self.revenue_event)
 
-    print(datetime.now(), 'finished loading CSV for', date.strftime('%d.%m.%Y'),
-          'took', datetime.now() - now)
+        fs = s3fs.S3FileSystem(anon=False)
+        fs.connect_timeout = 10  # defaults to 5
```

Comment:
Looks like better to set constants for both timeout values.

Reply:
It's a parameter we are not going to reconfigure pretty much ever, used once in the code => IMHO doesn't deserve a named constant